### PR TITLE
RequireFallback function for requiring userland code

### PR DESCRIPTION
**Pull request by nickmaltsev**
#1795

### CODE REVIEW

File: packages/gatsby/cache-dir/static-entry.js [hunk 1]
```diff
@@ -5,13 +5,9 @@ import { kebabCase, get, merge, isArray, isString } from "lodash"
...
 import syncRequires from "./sync-requires"
+import requireWithFallback from './require-with-fallback'
 
-let Html
-try {
-  Html = require(`../src/html`)
-} catch (e) {
-  Html = require(`./default-html`)
-}
+const Html = requireWithFallback(`../src/html`, `./default-html`)
```
Comment (jquense) [inline packages/gatsby/cache-dir/static-entry.js]: I don't think you can abstract this actually. Webpack statically analyzes requires, if the require path is completely dynamic the chances of stuff going wrong are high. like it including ALL files in the root in the bundle

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit e9bd3868b3f6c634907f23e285fa115e4386793e

https://app.netlify.com/sites/using-glamor/deploys/5991f8156f4c5071551d7008

File: packages/gatsby/cache-dir/static-entry.js (hunk 1 above)
Comment (nickmaltsev) [inline packages/gatsby/cache-dir/static-entry.js]: Didn't know about these webpack `features`. So, seems like the only way here is to add `e.code !== MODULE_NOT_FOUND` checking for each `try/catch` block?

Comment (jquense) [inline packages/gatsby/cache-dir/static-entry.js]: for any require in `cache-dir` yes, you can rescope the helper to contain the check and throw if you want though. but you need to do the require with string literals

**Comment by KyleAMathews** [comment]
Working on a PR for this exact issue right now actually :-)

And yeah, as @jquense said, you need to require the fallback module in the catch.

Also checking for `MODULE_NOT_FOUND` won't work I've learned as the module you're requiring requires a non-existent module then the code will still be `MODULE_NOT_FOUND`. Turns out the only foolproof way is do a regex check of the error string 🤷‍♂️

**Comment by KyleAMathews** [comment]
Thanks for looking into this! Will be reviewing your other PRs soon!

**Pull request closed without merging**